src/include/platform/NetworkCommissioning.h

```diff
@@ -0,0 +1,189 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+/**
+ *    @file
+ *      This file defines the CHIP Device Network Provisioning object.
+ *
+ */
+
+#pragma once
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <lib/core/CHIPCore.h>
+#include <lib/core/CHIPSafeCasts.h>
+#include <lib/support/ThreadOperationalDataset.h>
+#include <platform/internal/DeviceNetworkInfo.h>
+
+namespace chip {
+namespace DeviceLayer {
+
+class DLL_EXPORT WiFiNetworkCommissioningDelegate
+{
+public:
+    struct WiFiNetworkInfo
+    {
+        uint8_t mSSID[DeviceLayer::Internal::kMaxWiFiSSIDLength + 1];
+        uint8_t mSSIDLen;
+        uint8_t mCredentials[DeviceLayer::Internal::kMaxWiFiKeyLength];
+        uint8_t mCredentialsLen;
+    };
+
+    class Callback
+    {
+    public:
+        virtual void OnConnectWiFiResult(app::Clusters::NetworkCommissioning::NetworkCommissioningStatus commissioningError,
+                                         CharSpan debugText, int32_t wifiAssociationStatus) = 0;
+
+        /**
+         * Indicates the scan is finished, and accepts a list of networks discovered.
+         * OnScanFinished() must be called in a thread-safe manner with CHIP stack. (e.g. using ScheduleWork or ScheduleLambda)
+         */
+        virtual void
+        OnScanFinished(CHIP_ERROR err, CharSpan debugText,
+                       const Span<app::Clusters::NetworkCommissioning::Structs::WiFiInterfaceScanResult::Type> & network) = 0;
```

Comment:
Rename to `network` to `networks` (plural).

Would it be easier/better to have an iterator, rather than needing to maintain a **linear** span of large network results? The iterator would allow internal storage in any format (including native of driver).

Also, there is here again a dependency on generated code structs, which couples the view (cluster) to the model (driver). Strongly suggest defining a `WiFiScanResult` interface that only has accessors.  The underlying driver will be able to convert the results during iteration.

If the goal is to simplify the lifecycle of access to the data, the problem is not solved here. For instance, there is no guarantee that this OnScanFinished will be called on the correct context/thread to allow sending a response from the cluster. If you need to "forward" the callback to the correct context, even the current approach with a span requires the lifetime of the span to be longer than the lifetime of the callback, unless we fill the response data in callback context, but do the actual invoke response in a deferred way.

Overall, strongly suggest reconsidering the approach here to:

1. Avoid coupling to the exact cluster representation (it can be equivalent, but it must not be the generated code object)
2. Consider using an iterator pointer to a WiFiScanResult object type that only has getters, so that the callback impl can just iterate over the results and encode them as it sees fit, and let the the underlying driver convert between the internal storage representation and the current value being accessed/encoded
3. Look at GroupDataProviderImpl's iterators for an example of a simple iterator pattern that does the above (and does not depend on generated code) and that has safe handling of a `Release()` method for the iterator for when you are done with it.